src/network/NetworkManagerSecondary.cc

```diff
@@ -96,86 +113,129 @@ bool NetworkManagerSecondary::OnControl(const private_msgs::PeerControl &_req,
+      lastUpdateInfo.simTime = simTime;
+
+      // Throttle the number of step messages going to the debug output.
+      if (!info.paused && info.iterations % 1000 == 0)
+      {
+        igndbg << "Network iterations: " << info.iterations
+              << std::endl;
+      }
 
-      ignmsg << "Secondary [" << this->Namespace()
-             << "] assigned affinity to performer [" << entityId << "]."
-             << std::endl;
-    }
-    // If performer has been assigned to another secondary, remove it
-    else
-    {
-      auto parent =
-          this->dataPtr->ecm->Component<components::ParentEntity>(entityId);
-      this->dataPtr->ecm->RequestRemoveEntity(parent->Data());
-
-      if (this->performers.find(entityId) != this->performers.end())
+      // Update affinities
+      // TODO(ivanpauno):
+      // This needs to be redone based on if performers can "far apart", "can view each other" or "are interacting".
+      for (int i = 0; i < next_step.affinity_size(); ++i)
       {
-        ignmsg << "Secondary [" << this->Namespace()
-               << "] unassigned affinity to performer [" << entityId << "]."
-               << std::endl;
-        this->performers.erase(entityId);
+        const auto &affinityMsg = next_step.affinity(i);
+        const auto &entityId = affinityMsg.entity().id();
+
+        if (affinityMsg.secondary_prefix() == this->Namespace())
+        {
+          this->performers.insert(entityId);
+
+          ignmsg << "Secondary [" << this->Namespace()
+                << "] assigned affinity to performer [" << entityId << "]."
+                << std::endl;
+        }
+        // If performer has been assigned to another secondary, remove it
+        else
+        {
+          auto parent =
+              this->dataPtr->ecm->Component<components::ParentEntity>(entityId);
+          if (parent) {
+            this->dataPtr->ecm->RequestRemoveEntity(parent->Data());
+          }
+
+          if (this->performers.find(entityId) != this->performers.end())
+          {
+            ignmsg << "Secondary [" << this->Namespace()
+                  << "] unassigned affinity to performer [" << entityId << "]."
+                  << std::endl;
+            this->performers.erase(entityId);
+          }
+        }
       }
     }
-  }
-
-  // Update info
-  auto info = convert<UpdateInfo>(_msg.stats());
+    while (lastUpdateInfo.iterations < maxIteration) {
+      if (this->stopAsyncStepThread) {
+        return;
+      }
+      this->dataPtr->stepFunction(lastUpdateInfo);
 
-  // Step runner
-  this->dataPtr->stepFunction(info);
+      // Update state with all the performer's entities
+      std::unordered_set<Entity> entities;
+      for (const auto &perf : this->performers)
+      {
+        // Performer model
+        auto parent = this->dataPtr->ecm->Component<components::ParentEntity>(perf);
+        if (parent == nullptr)  // TODO(ivanpauno): why is this now needed?
```

Comment:
I think this isn't needed, if this check is triggered something really bad happened because the this->performers is updated right above.

Reply:
Good point, I believe that it can be removed.